src/Pos/Client/Txp/Util.hs

```diff
@@ -206,60 +261,215 @@ prepareInpsOuts utxo addrs outputs = do
-        uncurry (makePubKeyTx ss) <$>
-            prepareInpOuts
-                utxo
-                (makePubKeyAddress $ safeToPublic ss)
-                properOutputs
+    -> AddrData m
+    -> m (Either TxError TxWithSpendings)
+createTx utxo ss outputs addrData = runExceptT $
+    createGenericTxSingle (makePubKeyTx ss)
+    utxo outputs addrData
 
 -- | Make a transaction, using M-of-N script as a source
-createMOfNTx :: Utxo -> [(PublicKey, Maybe SafeSigner)] -> TxOutputs -> Either Text TxAux
-createMOfNTx utxo keys outputs = uncurry (makeMOfNTx validator sks) <$> inpOuts
-  where pks = map fst keys
-        sks = map snd keys
-        m = length $ filter isJust sks
-        validator = multisigValidator m pks
-        addr = makeScriptAddress validator
-        inpOuts = prepareInpOuts utxo addr outputs
-
-createRedemptionTx :: Utxo -> RedeemSecretKey -> TxOutputs -> Either Text TxAux
-createRedemptionTx utxo rsk outputs =
-    uncurry (makeRedemptionTx rsk) <$>
-    prepareInpOuts utxo (makeRedeemAddress $ redeemToPublic rsk) outputs
+createMOfNTx
+    :: TxCreateMode ctx m
+    => Utxo
+    -> [(PublicKey, Maybe SafeSigner)]
+    -> TxOutputs
+    -> AddrData m
+    -> m (Either TxError TxWithSpendings)
+createMOfNTx utxo keys outputs addrData = runExceptT $
+    createGenericTxSingle (makeMOfNTx validator sks)
+    utxo outputs addrData
+  where
+    pks = map fst keys
+    sks = map snd keys
+    m = length $ filter isJust sks
+    validator = multisigValidator m pks
+
+-- | Make a transaction for retrieving money from redemption address
+createRedemptionTx
+    :: TxDistrMode ctx m
+    => Utxo
+    -> RedeemSecretKey
+    -> TxOutputs
+    -> m (Either TxError TxAux)
+createRedemptionTx utxo rsk outputs = runExceptT $ do
+    TxRaw {..} <- prepareTxRaw utxo outputs (TxFee $ mkCoin 0)
+    let bareInputs = snd <$> trInputs
+    properOutputs <- overrideTxDistrBoot trOutputs
+    pure $ makeRedemptionTx rsk bareInputs properOutputs
+
+-----------------------------------------------------------------------------
+-- Fees logic
+-----------------------------------------------------------------------------
+
+-- | Helper function to reduce code duplication
+withLinearFeePolicy
+    :: (TxDistrMode ctx m, MonadError TxError m)
+    => (TxSizeLinear -> m a)
+    -> m a
+withLinearFeePolicy action = bvdTxFeePolicy <$> gsAdoptedBVData >>= \case
+    TxFeePolicyUnknown w _ -> throwTxError $
+        sformat ("Unknown fee policy, tag: "%build) w
+    TxFeePolicyTxSizeLinear linearPolicy ->
+        action linearPolicy
+
+-- | Prepare transaction considering fees
+prepareTxWithFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxRaw
+prepareTxWithFee utxo outputs = withLinearFeePolicy $ \linearPolicy ->
+    stabilizeTxFee linearPolicy utxo outputs
+
+-- | Compute, how much fees we should pay to send money to given
+-- outputs
+computeTxFee
+    :: TxDistrMode ctx m
```

Comment:
it's too general constraint for this fucntion. AFAIU it would be enough `MonadError`